pkg/reconciler/common/prune.go

```diff
@@ -0,0 +1,151 @@
+	if err != nil {
+		return err
+	}
+	if err := createCronJob(k, ctx, pru, targetNamespace, pruningNamespaces); err != nil {
+		logger.Error("failed to create cronjob ", err)
+	}
+
+	return nil
+}
+
+func GetPrunableNamespaces(k kubernetes.Interface, ctx context.Context) ([]string, error) {
+	nsList, err := k.CoreV1().Namespaces().List(ctx, v1.ListOptions{})
+	if err != nil {
+		return nil, err
+	}
+
+	var allNameSpaces []string
+	for _, ns := range nsList.Items {
+		if !strings.HasPrefix(ns.Name, openshiftPrefix) && !strings.HasPrefix(ns.Name, kubePrefix) {
+			allNameSpaces = append(allNameSpaces, ns.Name)
+		}
+	}
+	return allNameSpaces, nil
+}
+
+func createCronJob(k kubernetes.Interface, ctx context.Context, pru v1alpha1.Prune, targetNs string, pruningNs []string) error {
+	pruneContainers := getPruningContainers(pru.Resources, pruningNs, pru.Keep)
+	backOffLimit := int32(3)
+	ttlSeconfsAfterFindished := int32(3600)
+	cj := &v1beta1.CronJob{
+		TypeMeta: v1.TypeMeta{
+			Kind:       "CronJob",
+			APIVersion: "batch/v1beta1",
+		},
+		ObjectMeta: v1.ObjectMeta{
+			Name: CronName,
+		},
+		Spec: v1beta1.CronJobSpec{
+			Schedule:          pru.Schedule,
+			ConcurrencyPolicy: "Forbid",
+			JobTemplate: v1beta1.JobTemplateSpec{
+
+				Spec: batchv1.JobSpec{
+					TTLSecondsAfterFinished: &ttlSeconfsAfterFindished,
+					BackoffLimit:            &backOffLimit,
+
+					Template: corev1.PodTemplateSpec{
+						Spec: corev1.PodSpec{
+							Containers:         pruneContainers,
+							RestartPolicy:      "OnFailure",
+							ServiceAccountName: tektonSA,
+						},
+					},
+				},
+			},
+		},
+	}
+
+	if _, err := k.BatchV1beta1().CronJobs(targetNs).Create(ctx, cj, v1.CreateOptions{}); err != nil {
+		return err
+	}
+	return nil
+}
+
+func getPruningContainers(resources, namespaces []string, keep int) []corev1.Container {
+	containers := []corev1.Container{}
+	tknImage := getTknImage()
+
+	for _, res := range resources {
+		for _, ns := range namespaces {
+			cmdArgs := deleteCommand(res, keep, ns)
+			cName := res + "-pruner-tkn-" + ns
+			container := corev1.Container{
+				Name:                     cName,
+				Image:                    tknImage,
+				Command:                  []string{"tkn"},
+				Args:                     cmdArgs,
+				TerminationMessagePolicy: "FallbackToLogsOnError",
+			}
+			containers = append(containers, container)
+		}
+	}
+
+	return containers
+}
+
+func deleteCommand(res string, keep int, ns string) []string {
+	cmdArgs := []string{}
+	cmdArgs = append(cmdArgs, strings.ToLower(res))
+	cmdArgs = append(cmdArgs, "delete")
+	cmdArgs = append(cmdArgs, "--keep")
+	cmdArgs = append(cmdArgs, fmt.Sprint(keep))
+	cmdArgs = append(cmdArgs, "-f")
+	cmdArgs = append(cmdArgs, "-n")
+	cmdArgs = append(cmdArgs, ns)
+	return cmdArgs
+}
+
+func getTknImage() string {
+	imageFronEnv := os.Getenv(JobsTKNImageName)
+	if imageFronEnv != "" {
+		return imageFronEnv
+	}
+	return defaultTKNImage
+
+}
```

Comment:
Do we want to have a default or fail if the env is not there ?

Reply:
How about failing with an info message.